EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -319,47 +330,115 @@ int isContactorSet(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
 /*
   This function takes multiple reads because the current amplifier has a bit of a setup time
   to make sure that erroneous values aren't read, multiple reads are taken till we have a
-  relatively steady value. Dan suggested that 3 good reads in a row is good.
+  relatively steady value.
 */
 uint32_t readCurrentThroughContactors(void)
 {
     // Enable current amplifier. Allow time to settle and take multiple valid concurrent readings
     // to make sure it isn't changing
     HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_SET);
-    uint32_t sense = 0;
-    uint32_t prev_sense = 0;
+    uint32_t senses[NUM_SENSES];
+    int count = 0;
 
-    for (int counter = 0; counter < 3; counter++)
+    // Want to limit it so we don't get an infinite loop
+    while (count < MAX_CURRENT_READ_ATTEMPTS)
     {
+        resetSenseArray(senses);
         osDelay(SENSE_SETUP_TIME);
+        int ret_val = isSenseStable(senses);
 
-        if (HAL_ADC_PollForConversion(&hadc1, ADC_POLL_TIMEOUT) == HAL_OK)
+        if (ret_val == 1)
+        {
+            HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
+            return senses[0];
+        }
+        else if (ret_val == -1)
         {
-            sense = HAL_ADC_GetValue(&hadc1);
-            int diff = sense - prev_sense;
+            HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
+            return 0xDEADBEEF;
+        }
 
-            if (diff < 0)
-            {
-                diff *= -1;
-            }
+        count++;
+    }
 
-            if (diff > 50)
-            {
-                // sense is still not stable
-                counter = -1; // Will be incremented to 0 at the end of the loop
-            }
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
+    return 0xDEADBEEF;
+}
+
+/*
+  Multiple reads are taken and buffered into an array. The maximum and minimum value read are
+  found. Since these values will have the greatest difference, their difference must be less than
+  or equal to the allowable difference. If it is, the CURRENT_SENSE is stable, otherwise, it is deemed
+  not stable
+*/
+int isSenseStable(uint32_t senses[NUM_SENSES])
+{
+    for (int counter = 0; counter < NUM_SENSES; counter++)
+    {
+        for (int i = NUM_SENSES - 1; i > 0; i++)
+        {
+            senses[i] = senses[i - 1];
+        }
 
-            prev_sense = sense;
+        if (HAL_ADC_PollForConversion(&hadc1, ADC_POLL_TIMEOUT) == HAL_OK)
+        {
+            senses[0] = HAL_ADC_GetValue(&hadc1);
         }
         else
         {
-            sense = 0xDEADBEEF;
-            break;
+            return -1;
         }
     }
 
-    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_RESET);
-    return sense;
+    uint32_t max = findMax(senses);
+    uint32_t min = findMin(senses);
```

Comment:
I think the two funcitons findMax & findMin can be combined, to reduce the amount of looping going on:
````
    for (int i = 1; i < NUM_SENSES; i++)
    {
        uint32_t max = senses[0];
        uint32_t min = senses[0];
        if (senses[i] < min)
        {
            min = senses[i];
        }
  if (senses[] > max)
        {
            max = senses[i];
        }
    }
````